log4j-appender/src/main/java/org/apache/kafka/log4jappender/KafkaLog4jAppender.java

```diff
@@ -349,8 +349,14 @@ protected Producer<byte[], byte[]> getKafkaProducer(Properties props) {
     protected void append(LoggingEvent event) {
         String message = subAppend(event);
         LogLog.debug("[" + new Date(event.getTimeStamp()) + "]" + message);
-        Future<RecordMetadata> response = producer.send(
-            new ProducerRecord<>(topic, message.getBytes(StandardCharsets.UTF_8)));
+        Future<RecordMetadata> response;
```

Comment:
nice finding. one question: should this function log message instead of sending data by closed producer if `closed` is true? for example:
```java
        if (closed) {
            LogLog.debug("sending to kafka by closed producer");
            return;
        }
```

Reply:
I don't think so. The internal state of the producer is the source of truth on whether this is closed, not the `closed` variable. A race condition could still cause an IllegalStateException to occur even if there's an if-condition first.

Reply:
> I don't think so. The internal state of the producer is the source of truth on whether this is closed, not the closed variable. 

That makes sense. one more question: Is `IllegalStateException` caused only by "closed" producer?

Reply:
This is in the javadoc for send:
```
     * @throws IllegalStateException if a transactional.id has been configured and no transaction has been started, or
     *                               when send is invoked after producer has been closed.
```
Idempotence and transactions are not in use for this producer. I am not aware of another cause of IllegalStateException that isn't documented, but the documentation could be wrong or there could be a bug, I have no idea.